### DISCUSSION TITLE
How to use scons program with custom configuration?

### DISCUSSION THREAD

**Discussion by the-risk-taker** [modm-io/modm#867] [Q&A] [answered]
Hi,

I have problem with running `scons program` on custom configuration. I have two projects, one of which extends the NUCLEO-F103RB:

```
<library>
  <extends>modm:nucleo-f103rb</extends>
  <options>
    <option name="modm:build:cmake:include_cmakelists">False</option> <!--Change to True to generate first CMakeLists.txt file-->
    <option name="modm:build:scons:include_sconstruct">False</option> <!--Change to True to generate simple SConstruct file-->
  </options>
  <modules>
    <module>modm:build:cmake</module>
    <module>modm:build:scons</module>
    <module>modm:processing:timer</module>
    <module>modm:platform:timer:2</module>
  </modules>
  <repositories>
    <repository><path>../submodules/modm/repo.lb</path></repository>
  </repositories>
</library>
```

Where running `lbuild build` and `scons program` works fine - the program is flashed on the board.

Then I have a custom configuration targeting the same MCU as on NUCLEO-F103RB:

```
<library>
   <options>
      <option name=":target">stm32f103rbt6</option>
      <option name=":build:build.path">build</option>
      <option name="modm:build:cmake:include_cmakelists">False</option> <!--Change to True to generate first CMakeLists.txt file-->
      <option name="modm:build:scons:include_sconstruct">False</option> <!--Change to True to generate simple SConstruct file-->
   </options>
   <modules>
      <module>modm:build:scons</module>
      <module>modm:build:cmake</module>
      <module>modm:docs</module>
      <module>modm:debug</module>
      <module>modm:architecture:clock</module>
      <module>modm:platform:clock</module>
      <module>modm:platform:core</module>
      <module>modm:platform:gpio</module>
      <module>modm:platform:uart:2</module>
   </modules>
   <repositories>
      <repository><path>../submodules/modm/repo.lb</path></repository>
   </repositories>
</library>
```

Here, I can build program, but the `scons program` does not work, it's saying that:

```
scons: Building targets ...
╭────────────── build-scons/scons-release/modm-example-target.elf
╰───OpenOCD───> stm32f103rbt6
Open On-Chip Debugger 0.11.0+dev-00687-g7d2ea186c (2022-05-20-12:38)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Error: Debug Adapter has to be specified, see "adapter driver" command
```

What am I missing, what do I need to do, to make it work?

**Comment by salkinium** [comment, marked as answer]
You need to tell OpenOCD how to flash you custom board, ie. what adapter is used, what Reset config, what MCU, inside a `openocd.cfg` file, and then specify it via the [`<option name="modm:build:openocd.cfg">openocd.cfg</option>` path option](https://modm.io/reference/module/modm-build/#openocdcfg).
All [the F103 BluePill example do it this way](https://github.com/modm-io/modm/tree/develop/examples/blue_pill_f103/blink) to tell OpenOCD what Debug Adapter to use.

Alternatively, if your config already exists inside OpenOCD support, you can add it via [the `openocd.source` collector](https://modm.io/reference/module/modm-build/#openocdsource): `<collect name="modm:build:openocd.source">board/st_nucleo_f103rb.cfg</collect>`, which is what [the BSP does for you internally](https://github.com/modm-io/modm/blob/develop/src/modm/board/nucleo_f103rb/module.lb#L39).

**Reply by the-risk-taker** [reply]
Now it works as I wanted.  Thank you very much for quick response 👍

For future readers: I added file `openocd.cfg` (in the same dir as `project.xml`), then in  `openocd.cfg` I put:

```
source [find interface/stlink.cfg]
source [find target/stm32f1x.cfg]
transport select hla_swd
reset_config none
```

Next in the `project.xml`, I added:

```
   <options>
      ...
      <option name="modm:build:openocd.cfg">openocd.cfg</option>
   </options>
```
